// @storybook/addon-actions/dist/index.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "action", {
  enumerable: true,
  get: function get() {
    return _preview.action;
  }
});
Object.defineProperty(exports, "actions", {
  enumerable: true,
  get: function get() {
    return _preview.actions;
  }
});
Object.defineProperty(exports, "decorate", {
  enumerable: true,
  get: function get() {
    return _preview.decorate;
  }
});
Object.defineProperty(exports, "configureActions", {
  enumerable: true,
  get: function get() {
    return _preview.configureActions;
  }
});
Object.defineProperty(exports, "decorateAction", {
  enumerable: true,
  get: function get() {
    return _preview.decorateAction;
  }
});
Object.defineProperty(exports, "withActions", {
  enumerable: true,
  get: function get() {
    return _preview.withActions;
  }
});
Object.defineProperty(exports, "ADDON_ID", {
  enumerable: true,
  get: function get() {
    return _constants.ADDON_ID;
  }
});
Object.defineProperty(exports, "PANEL_ID", {
  enumerable: true,
  get: function get() {
    return _constants.PANEL_ID;
  }
});
Object.defineProperty(exports, "EVENT_ID", {
  enumerable: true,
  get: function get() {
    return _constants.EVENT_ID;
  }
});

var _preview = require("./preview");

var _constants = require("./constants");
