// @storybook/addon-a11y/dist/components/Report/Elements.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("core-js/modules/es6.string.iterator");

require("core-js/modules/es6.array.from");

require("core-js/modules/es6.regexp.to-string");

require("core-js/modules/es6.date.to-string");

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.is-array");

require("core-js/modules/es6.array.map");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _Rules = _interopRequireDefault(require("./Rules"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var Item =
/*#__PURE__*/
(0, _styled.default)("li", {
  target: "eji8sqs0"
})({
  fontWeight: 600
});
var ItemTitle =
/*#__PURE__*/
(0, _styled.default)("span", {
  target: "eji8sqs1"
})({
  borderBottom: '1px solid rgb(130, 130, 130)',
  width: '100%',
  display: 'inline-block',
  paddingBottom: '4px',
  marginBottom: '4px'
});

function Element(_ref) {
  var element = _ref.element,
      passes = _ref.passes;
  var any = element.any,
      all = element.all,
      none = element.none;

  var rules = _toConsumableArray(any).concat(_toConsumableArray(all), _toConsumableArray(none));

  return _react.default.createElement(Item, null, _react.default.createElement(ItemTitle, null, element.target[0]), _react.default.createElement(_Rules.default, {
    rules: rules,
    passes: passes
  }));
}

Element.propTypes = {
  element: _propTypes.default.shape({
    any: _propTypes.default.array.isRequired,
    all: _propTypes.default.array.isRequired,
    none: _propTypes.default.array.isRequired
  }).isRequired,
  passes: _propTypes.default.bool.isRequired
};
/* eslint-disable react/no-array-index-key */

var Elements = function Elements(_ref2) {
  var elements = _ref2.elements,
      passes = _ref2.passes;
  return _react.default.createElement("ol", null, elements.map(function (element, index) {
    return _react.default.createElement(Element, {
      passes: passes,
      element: element,
      key: index
    });
  }));
};

Elements.propTypes = {
  elements: _propTypes.default.arrayOf(_propTypes.default.shape({
    any: _propTypes.default.array.isRequired,
    all: _propTypes.default.array.isRequired,
    none: _propTypes.default.array.isRequired
  })).isRequired,
  passes: _propTypes.default.bool.isRequired
};
var _default = Elements;
exports.default = _default;
