// @storybook/addon-a11y/dist/components/Report/Info.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Wrapper =
/*#__PURE__*/
(0, _styled.default)("div", {
  target: "e1uet5c10"
})(function (_ref) {
  var theme = _ref.theme;
  return {
    backgroundColor: theme.barFill,
    padding: '12px',
    marginBottom: '10px'
  };
});
var Help =
/*#__PURE__*/
(0, _styled.default)("p", {
  target: "e1uet5c11"
})({
  margin: '0 0 12px'
});
var Link =
/*#__PURE__*/
(0, _styled.default)("a", {
  target: "e1uet5c12"
})({
  marginTop: '12px',
  textDecoration: 'underline',
  color: 'inherit',
  display: 'block'
});

function Info(_ref2) {
  var item = _ref2.item;
  return _react.default.createElement(Wrapper, null, _react.default.createElement(Help, null, item.help), _react.default.createElement(Link, {
    href: item.helpUrl,
    target: "_blank"
  }, "More info..."));
}

Info.propTypes = {
  item: _propTypes.default.shape({
    help: _propTypes.default.node,
    helpUrl: _propTypes.default.string
  }).isRequired
};
var _default = Info;
exports.default = _default;
