// @storybook/addon-a11y/dist/components/Report/Tags.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("core-js/modules/es6.array.map");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Wrapper =
/*#__PURE__*/
(0, _styled.default)("div", {
  target: "ez33kbu0"
})({
  display: 'flex',
  flexWrap: 'wrap',
  margin: '12px 0'
});
var Item =
/*#__PURE__*/
(0, _styled.default)("div", {
  target: "ez33kbu1"
})(function (_ref) {
  var theme = _ref.theme;
  return {
    margin: '0 6px',
    padding: '5px',
    border: theme.mainBorder,
    borderRadius: theme.mainBorderRadius
  };
});

function Tags(_ref2) {
  var tags = _ref2.tags;
  return _react.default.createElement(Wrapper, null, tags.map(function (tag) {
    return _react.default.createElement(Item, {
      key: tag
    }, tag);
  }));
}

Tags.propTypes = {
  tags: _propTypes.default.arrayOf(_propTypes.default.node).isRequired
};
var _default = Tags;
exports.default = _default;
