// @storybook/ui/dist/libs/key_events.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isModifierPressed = isModifierPressed;
exports.default = handle;
exports.handleKeyboardShortcuts = handleKeyboardShortcuts;
exports.features = void 0;

var _keycode = _interopRequireDefault(require("keycode"));

var _coreEvents = _interopRequireDefault(require("@storybook/core-events"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var features = {
  FULLSCREEN: 'FULLSCREEN',
  ADDON_PANEL: 'ADDON_PANEL',
  STORIES_PANEL: 'STORIES_PANEL',
  SHORTCUTS_HELP: 'SHORTCUTS_HELP',
  ESCAPE: 'ESCAPE',
  NEXT_STORY: 'NEXT_STORY',
  PREV_STORY: 'PREV_STORY',
  SHOW_SEARCH: 'SHOW_SEARCH',
  ADDON_PANEL_IN_RIGHT: 'ADDON_PANEL_IN_RIGHT'
};
exports.features = features;

function isModifierPressed(e) {
  return (e.ctrlKey || e.keyCode === 91 || e.metaKey) && e.shiftKey;
}

function focusInInput(e) {
  return /input|textarea/i.test(e.target.tagName) || e.target.getAttribute('contenteditable') !== null;
}

function handle(e) {
  if (e.keyCode === (0, _keycode.default)('escape')) {
    // We don't need to preventDefault escape.
    // Just getting the event is enough for us.
    return features.ESCAPE;
  }

  if (focusInInput(e)) {
    // if we're focused in an element that accepts input,
    // then we shouldn't perform a shortcut action
    return false;
  }

  if (!isModifierPressed(e)) return false;

  switch (e.keyCode) {
    case (0, _keycode.default)('F'):
      e.preventDefault();
      return features.FULLSCREEN;

    case (0, _keycode.default)('Z'):
      e.preventDefault();
      return features.ADDON_PANEL;

    case (0, _keycode.default)('X'):
      e.preventDefault();
      return features.STORIES_PANEL;

    case (0, _keycode.default)('right'):
      e.preventDefault();
      return features.NEXT_STORY;

    case (0, _keycode.default)('left'):
      e.preventDefault();
      return features.PREV_STORY;

    case (0, _keycode.default)('O'):
      e.preventDefault();
      return features.SHOW_SEARCH;

    case (0, _keycode.default)('G'):
      e.preventDefault();
      return features.ADDON_PANEL_IN_RIGHT;

    default:
      return false;
  }
} // window.keydown handler to dispatch a key event to the preview channel


function handleKeyboardShortcuts(channel) {
  return function (event) {
    var parsedEvent = handle(event);

    if (parsedEvent) {
      channel.emit(_coreEvents.default.APPLY_SHORTCUT, {
        event: parsedEvent
      });
    }
  };
}
