// @storybook/addon-notes/dist/register.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Notes = void 0;

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/es6.object.define-property");

require("core-js/modules/es6.object.create");

require("core-js/modules/es6.object.set-prototype-of");

require("core-js/modules/es6.string.trim");

require("core-js/modules/es6.regexp.replace");

require("core-js/modules/es6.function.bind");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

var Panel =
/*#__PURE__*/
(0, _styled.default)("div", {
  target: "eap90390"
})({
  padding: 10,
  boxSizing: 'border-box',
  width: '100%'
});

var Notes =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Notes, _React$Component);

  function Notes() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Notes);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Notes)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      text: ''
    };
    _this.onAddNotes = _this.onAddNotes.bind(_assertThisInitialized(_assertThisInitialized(_this)));
    return _this;
  }

  _createClass(Notes, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      var _this$props = this.props,
          channel = _this$props.channel,
          api = _this$props.api; // Listen to the notes and render it.

      channel.on('storybook/notes/add_notes', this.onAddNotes); // Clear the current notes on every story change.

      this.stopListeningOnStory = api.onStory(function () {
        _this2.onAddNotes('');
      });
    } // This is some cleanup tasks when the Notes panel is unmounting.

  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.stopListeningOnStory) {
        this.stopListeningOnStory();
      }

      this.unmounted = true;
      var channel = this.props.channel;
      channel.removeListener('storybook/notes/add_notes', this.onAddNotes);
    }
  }, {
    key: "onAddNotes",
    value: function onAddNotes(text) {
      this.setState({
        text: text
      });
    }
  }, {
    key: "render",
    value: function render() {
      var active = this.props.active;
      var text = this.state.text;
      var textAfterFormatted = text ? text.trim().replace(/(<\S+.*>)\n/g, '$1').replace(/\n/g, '<br />') : '';
      return active ? _react.default.createElement(Panel, {
        className: "addon-notes-container",
        dangerouslySetInnerHTML: {
          __html: textAfterFormatted
        }
      }) : null;
    }
  }]);

  return Notes;
}(_react.default.Component);

exports.Notes = Notes;
Notes.propTypes = {
  active: _propTypes.default.bool.isRequired,
  channel: _propTypes.default.shape({
    on: _propTypes.default.func,
    emit: _propTypes.default.func,
    removeListener: _propTypes.default.func
  }).isRequired,
  api: _propTypes.default.shape({
    onStory: _propTypes.default.func,
    getQueryParam: _propTypes.default.func,
    setQueryParams: _propTypes.default.func
  }).isRequired
};

_addons.default.register('storybook/notes', function (api) {
  var channel = _addons.default.getChannel();

  _addons.default.addPanel('storybook/notes/panel', {
    title: 'Notes',
    // eslint-disable-next-line react/prop-types
    render: function render(_ref) {
      var active = _ref.active;
      return _react.default.createElement(Notes, {
        channel: channel,
        api: api,
        active: active
      });
    }
  });
});
