// @storybook/node-logger/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.logger = exports.colors = void 0;

var _npmlog = _interopRequireDefault(require("npmlog"));

var _prettyHrtime = _interopRequireDefault(require("pretty-hrtime"));

var _chalk = _interopRequireDefault(require("chalk"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const colors = {
  pink: _chalk.default.hex('F1618C'),
  purple: _chalk.default.hex('B57EE5'),
  orange: _chalk.default.hex('F3AD38'),
  green: _chalk.default.hex('A2E05E'),
  blue: _chalk.default.hex('6DABF5'),
  red: _chalk.default.hex('F16161'),
  gray: _chalk.default.gray
};
exports.colors = colors;
const logger = {
  info: message => _npmlog.default.info('', message),
  warn: message => _npmlog.default.warn('', message),
  error: message => _npmlog.default.error('', message),
  trace: ({
    message,
    time
  }) => _npmlog.default.info(`${message} (${colors.purple((0, _prettyHrtime.default)(time))})`)
};
exports.logger = logger;
