// @storybook/core/dist/server/manager/manager-webpack.config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _webpack = _interopRequireDefault(require("webpack"));

var _dotenvWebpack = _interopRequireDefault(require("dotenv-webpack"));

var _htmlWebpackPlugin = _interopRequireDefault(require("html-webpack-plugin"));

var _caseSensitivePathsWebpackPlugin = _interopRequireDefault(require("case-sensitive-paths-webpack-plugin"));

var _paths = _interopRequireDefault(require("@storybook/ui/paths"));

var _findCacheDir = _interopRequireDefault(require("find-cache-dir"));

var _package = require("../../../package.json");

var _template = require("../utils/template");

var _utils = require("../config/utils");

var _babelLoader = _interopRequireDefault(require("../common/babel-loader"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const coreDirName = _path.default.dirname(require.resolve('@storybook/core/package.json'));

const context = _path.default.join(coreDirName, '../../node_modules');

const cacheDir = (0, _findCacheDir.default)({
  name: 'storybook'
});

var _default = ({
  configDir,
  configType,
  entries,
  outputDir,
  cache,
  babelOptions
}) => {
  const {
    raw,
    stringified
  } = (0, _utils.loadEnv)();
  const isProd = configType === 'PRODUCTION';
  return {
    name: 'manager',
    mode: isProd ? 'production' : 'development',
    bail: isProd,
    devtool: 'none',
    entry: entries,
    output: {
      path: outputDir,
      filename: '[name].[chunkhash].bundle.js',
      publicPath: ''
    },
    cache,
    plugins: [new _webpack.default.DllReferencePlugin({
      context,
      manifest: _path.default.join(__dirname, '../../../dll/storybook_ui-manifest.json')
    }), new _htmlWebpackPlugin.default({
      filename: `index.html`,
      chunksSortMode: 'none',
      alwaysWriteToDisk: true,
      inject: false,
      templateParameters: (compilation, files, options) => ({
        compilation,
        files,
        options,
        version: _package.version,
        dlls: ['./sb_dll/storybook_ui_dll.js'],
        headHtmlSnippet: (0, _template.getManagerHeadHtml)(configDir, process.env)
      }),
      template: require.resolve(`../templates/index.ejs`)
    }), new _webpack.default.DefinePlugin({
      'process.env': stringified
    }), new _caseSensitivePathsWebpackPlugin.default(), new _dotenvWebpack.default({
      silent: true
    })],
    module: {
      rules: [(0, _babelLoader.default)(babelOptions)]
    },
    resolve: {
      extensions: ['.mjs', '.js', '.jsx', '.json'],
      modules: ['node_modules'].concat(raw.NODE_PATH || []),
      alias: _objectSpread({
        'core-js': _path.default.dirname(require.resolve('core-js/package.json'))
      }, _paths.default)
    },
    recordsPath: _path.default.join(cacheDir, 'records.json'),
    optimization: {
      splitChunks: {
        chunks: 'all'
      },
      runtimeChunk: true
    }
  };
};

exports.default = _default;
