// @storybook/ui/dist/modules/ui/containers/header.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.mapper = void 0;

require("core-js/modules/es6.function.name");

var _pick = _interopRequireDefault(require("lodash/pick"));

var _components = require("@storybook/components");

var _gen_podda_loader = _interopRequireDefault(require("../libs/gen_podda_loader"));

var _compose = _interopRequireDefault(require("../../../compose"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mapper = function mapper(state, props, _ref) {
  var actions = _ref.actions;
  var currentOptions = (0, _pick.default)(state.shortcutOptions, 'showStoriesPanel', 'enableShortcuts', 'addonPanelInRight');
  var actionMap = actions();
  var uiOptions = state.uiOptions,
      isMobileDevice = state.isMobileDevice;
  var _uiOptions$name = uiOptions.name,
      name = _uiOptions$name === void 0 ? '' : _uiOptions$name,
      _uiOptions$url = uiOptions.url,
      url = _uiOptions$url === void 0 ? '' : _uiOptions$url;

  var handleBurgerButtonClick = function handleBurgerButtonClick() {
    actionMap.shortcuts.setOptions({
      showStoriesPanel: !currentOptions.showStoriesPanel
    });
  };

  var addonPanelInRight = isMobileDevice ? false : currentOptions.addonPanelInRight;
  return {
    name: name,
    url: url,
    enableShortcutsHelp: currentOptions.enableShortcuts,
    openShortcutsHelp: actionMap.ui.toggleShortcutsHelp,
    onBurgerButtonClick: handleBurgerButtonClick,
    isMobileDevice: isMobileDevice,
    addonPanelInRight: addonPanelInRight
  };
};

exports.mapper = mapper;

var _default = (0, _compose.default)((0, _gen_podda_loader.default)(mapper), {
  withRef: false
})(_components.Header);

exports.default = _default;
