// @storybook/ui/dist/modules/ui/containers/search_box.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.mapper = void 0;

var _search_box = _interopRequireDefault(require("../components/search_box"));

var _gen_podda_loader = _interopRequireDefault(require("../libs/gen_podda_loader"));

var _compose = _interopRequireDefault(require("../../../compose"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mapper = function mapper(state, props, _ref) {
  var actions = _ref.actions;
  var actionMap = actions();
  return {
    showSearchBox: state.shortcutOptions.showSearchBox,
    stories: state.stories,
    onSelectStory: actionMap.api.selectStory,
    onClose: function onClose() {
      actionMap.shortcuts.setOptions({
        showSearchBox: false
      });
    }
  };
};

exports.mapper = mapper;

var _default = (0, _compose.default)((0, _gen_podda_loader.default)(mapper), {
  withRef: false
})(_search_box.default);

exports.default = _default;
