// @storybook/ui/dist/modules/ui/components/search_box.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SearchBox = void 0;

require("core-js/modules/es6.object.assign");

require("core-js/modules/es7.symbol.async-iterator");

require("core-js/modules/es6.symbol");

require("core-js/modules/es6.object.define-property");

require("core-js/modules/es6.object.create");

require("core-js/modules/es6.object.set-prototype-of");

require("core-js/modules/es6.array.map");

require("core-js/modules/web.dom.iterable");

require("core-js/modules/es6.array.for-each");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _styled = _interopRequireDefault(require("@emotion/styled"));

var _reactLifecyclesCompat = require("react-lifecycles-compat");

var _reactModal = _interopRequireDefault(require("react-modal"));

var _reactFuzzy = _interopRequireDefault(require("react-fuzzy"));

var _core = require("@emotion/core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var formatStories = function formatStories(stories) {
  var formattedStories = [];
  var i = 0;
  stories.forEach(function (val) {
    i += 1;
    formattedStories.push({
      type: 'kind',
      value: val.kind,
      id: i
    });
    val.stories.forEach(function (story) {
      i += 1;
      formattedStories.push({
        type: 'story',
        value: story,
        id: i,
        kind: val.kind
      });
    });
  });
  return formattedStories;
};

var SuggestionWrapper =
/*#__PURE__*/
(0, _styled.default)("div", {
  target: "e1h5sroh0"
})(function (_ref) {
  var selected = _ref.selected,
      selectedResultStyle = _ref.selectedResultStyle,
      resultsStyle = _ref.resultsStyle;
  return selected ? selectedResultStyle : resultsStyle;
}, {
  display: 'flex',
  justifyContent: 'space-between'
});
var SuggestionValue =
/*#__PURE__*/
(0, _styled.default)("p", {
  target: "e1h5sroh1"
})({
  margin: 0
});
var SuggestionKind =
/*#__PURE__*/
(0, _styled.default)("p", {
  target: "e1h5sroh2"
})({
  opacity: 0.5,
  margin: 0,
  paddingLeft: 10,
  textAlign: 'right'
});

var suggestionTemplate = function suggestionTemplate(props, state, styles, clickHandler) {
  return state.results.map(function (val, i) {
    return _react.default.createElement(SuggestionWrapper, _extends({
      selected: state.selectedIndex === i
    }, styles, {
      tabIndex: 0,
      role: "option",
      "aria-selected": state.selectedIndex === i,
      key: "".concat(val.value, "_").concat(val.id),
      onClick: function onClick() {
        return clickHandler(i);
      }
    }), _react.default.createElement(SuggestionValue, null, val.value), _react.default.createElement(SuggestionKind, null, val.type === 'story' ? "in ".concat(val.kind) : 'Kind'));
  });
};

var SearchBox =
/*#__PURE__*/
function (_React$Component) {
  _inherits(SearchBox, _React$Component);

  function SearchBox() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, SearchBox);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(SearchBox)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onSelect", function (selected) {
      var onClose = _this.props.onClose;

      if (selected.type === 'story') {
        _this.fireOnStory(selected.value, selected.kind);
      } else {
        _this.fireOnKind(selected.value);
      }

      onClose();
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onModalOpen", function () {
      if (_this.inputRef.refs.searchBox !== null) {
        _this.inputRef.refs.searchBox.focus();
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "fireOnKind", function (kind) {
      var onSelectStory = _this.props.onSelectStory;

      if (onSelectStory) {
        onSelectStory(kind, null);
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "fireOnStory", function (story, kind) {
      var onSelectStory = _this.props.onSelectStory;

      if (onSelectStory) {
        onSelectStory(kind, story);
      }
    });

    return _this;
  }

  _createClass(SearchBox, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      var _this$props = this.props,
          showSearchBox = _this$props.showSearchBox,
          onClose = _this$props.onClose,
          stories = _this$props.stories,
          theme = _this$props.theme;
      return _react.default.createElement(_reactModal.default, {
        isOpen: showSearchBox,
        onAfterOpen: this.onModalOpen,
        onRequestClose: onClose,
        style: {
          overlay: {
            height: '100vh',
            width: '100vw',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            background: theme.overlayBackground,
            zIndex: 1
          },
          content: {
            fontSize: theme.mainTextColor,
            fontFamily: theme.mainTextFace,
            color: theme.mainTextColor
          }
        },
        className: "tmp" // This exists to make the Modal discard it's default styles for content
        ,
        contentLabel: "Search",
        shouldReturnFocusAfterClose: false
      }, _react.default.createElement(_reactFuzzy.default, {
        list: formatStories(stories),
        onSelect: this.onSelect,
        keys: ['value', 'type'],
        resultsTemplate: suggestionTemplate,
        ref: function ref(inputRef) {
          _this2.inputRef = inputRef;
        }
      }));
    }
  }]);

  return SearchBox;
}(_react.default.Component);

exports.SearchBox = SearchBox;
SearchBox.defaultProps = {
  stories: []
};
SearchBox.propTypes = {
  theme: _propTypes.default.shape({}).isRequired,
  showSearchBox: _propTypes.default.bool.isRequired,
  stories: _propTypes.default.arrayOf(_propTypes.default.object),
  onSelectStory: _propTypes.default.func.isRequired,
  onClose: _propTypes.default.func.isRequired
};
(0, _reactLifecyclesCompat.polyfill)(SearchBox);

var _default = (0, _core.withCSSContext)(function (props, _ref2) {
  var theme = _ref2.theme;
  return _react.default.createElement(SearchBox, _extends({}, props, {
    theme: theme
  }));
});

exports.default = _default;
