// @storybook/addon-actions/dist/preview/index.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "action", {
  enumerable: true,
  get: function get() {
    return _action.default;
  }
});
Object.defineProperty(exports, "actions", {
  enumerable: true,
  get: function get() {
    return _actions.default;
  }
});
Object.defineProperty(exports, "configureActions", {
  enumerable: true,
  get: function get() {
    return _configureActions.configureActions;
  }
});
Object.defineProperty(exports, "decorateAction", {
  enumerable: true,
  get: function get() {
    return _decorateAction.decorateAction;
  }
});
Object.defineProperty(exports, "decorate", {
  enumerable: true,
  get: function get() {
    return _decorateAction.decorate;
  }
});
Object.defineProperty(exports, "withActions", {
  enumerable: true,
  get: function get() {
    return _withActions.default;
  }
});

var _action = _interopRequireDefault(require("./action"));

var _actions = _interopRequireDefault(require("./actions"));

var _configureActions = require("./configureActions");

var _decorateAction = require("./decorateAction");

var _withActions = _interopRequireDefault(require("./withActions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
