// @storybook/ui/dist/modules/ui/routes.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _reactModal = _interopRequireDefault(require("react-modal"));

var _layout = _interopRequireDefault(require("./containers/layout"));

var _stories_panel = _interopRequireDefault(require("./containers/stories_panel"));

var _addon_panel = _interopRequireDefault(require("./containers/addon_panel"));

var _shortcuts_help = _interopRequireDefault(require("./containers/shortcuts_help"));

var _search_box = _interopRequireDefault(require("./containers/search_box"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _default(injectDeps, _ref) {
  var clientStore = _ref.clientStore,
      provider = _ref.provider,
      domNode = _ref.domNode;
  var state = clientStore.getAll(); // generate preview

  var Preview = function Preview() {
    var preview = provider.renderPreview(state.selectedKind, state.selectedStory);
    return preview;
  }; // Tell react-modal which element to mark as aria-hidden


  _reactModal.default.setAppElement(domNode);

  var Container = process.env.STORYBOOK_EXAMPLE_APP ? _react.default.StrictMode : 'div';

  var root = _react.default.createElement(Container, null, _react.default.createElement(_layout.default, {
    storiesPanel: function storiesPanel() {
      return _react.default.createElement(_stories_panel.default, null);
    },
    preview: function preview() {
      return _react.default.createElement(Preview, null);
    },
    addonPanel: function addonPanel() {
      return _react.default.createElement(_addon_panel.default, null);
    },
    shortcutsHelp: function shortcutsHelp() {
      return _react.default.createElement(_shortcuts_help.default, null);
    },
    searchBox: function searchBox() {
      return _react.default.createElement(_search_box.default, null);
    }
  }));

  _reactDom.default.render(root, domNode);
}
